CONTRIBUTING.md

```diff
@@ -0,0 +1,64 @@
+
+Pull requests should target the release/vX.Y branch instead of master. The latest code in RKE is included in such release branches.
+
+### Writing good commit messages
+Git commit messages should explain the how and why of your change and be separated into a brief subject line followed by a more detailed body. When in doubt, follow this guide for good commit messages and you can’t go wrong: https://chris.beams.io/posts/git-commit/.
+
+One particularly useful point made in the above guide is regarding commit subject lines:
+
+> A properly formed Git commit subject line should always be able to complete the following sentence:
+> 
+> - If applied, this commit will <ins>your subject line here</ins>
+
+A simple but effective convention to follow for commits is the “problem / solution” pattern. It looks like this:
+```
+<Subject>
+
+Problem: <Statement of problem>
+
+Solution: <Statement of solution>
+```
+
+As an example, here is a commit taken from the rancher/rancher repo:
+```
+commit b71ce2892eecb7c87a5212e3486f1de899a694aa
+Author: Dan Ramich <danold215@gmail.com>
+Date:   Tue Jun 19 11:56:52 2018 -0700
+
+    Add Validator for RoleTemplate
+
+    Problem:
+    Builtin RoleTemplates can be updated through the API
+
+    Solution:
+    Add a Validator to ensure the only field that can be changed on a
+    builtin RoleTemplate is 'locked'
+```
+
+### Reviewing, addressing feedback, and merging
+Generally, pull requests need two approvals from maintainers to be merged.
+
+When addressing review feedback, it is helpful to the reviewer if additional changes are made in new commits. This allows the reviewer to easily see the delta between what they previously reviewed and the changes you added to address their feedback.
+
+Once a PR has the necessary approvals, it can be merged. Here’s how the merge should be handled:
+- If the PR is a single logical commit, the merger should use the “Rebase and merge” option. This keeps the git commit history very clean and simple and eliminates noise from "merge commits."
+- If the PR is more than one logical commit, the merger should use the “Create a merge commit” option.
+- If the PR consists of more than one commit because the author added commits to address feedback, the commits should be squashed into a single commit (or more than one logical commit, if it is a big feature that needs more commits). This can be achieved in one of two ways:
+  - The merger can use the “Squash and merge” option. If they do this, the merger is responsible for cleaning up the commit message according to the previously stated commit message guidance.
+  - The pull request author, after getting the requisite approvals, can reorganize the commits as they see fit (using, for example, git rebase -i) and re-push.
+
+### Backports
```

Comment:
This is not enough context for a contributor. `A commit` can be core RKE code that applies to everything, which could be made k8s version (minor/patch) specific or only applies to a subset like an add-on which is already k8s version specific. This needs to be split up in the bigger currently known changes we have and address those.

Reply:
Good point but I don't get what you mean by `needs to be split up in the bigger currently known changes we have and address those`. I'll try to rewrite it to cover the different backport cases

Reply:
We know what kind of changes we have, these need to be described in a bit more detail so the reader can identify what change falls under what category. If the reader can do that, the reader can also identify where it needs to be backported.

We have RKE code backports, RKE k8s version backports and add-on templates backports at minimum which all have their own "policy" or at least it needs to be described.

Reply:
This is what I have written.
RKE code backports:
```
By design, RKE code supports all active Kubernetes minor versions, so the commit in the RKE project is automatically available for all. You just need to remember to contribute the commit in all the active RKE releases which are developed in `release vX.Y` branches
```
add-on templates backports:
```
if the commit is done in a dependent project, such as an add-on, the `data.json` file will need to be updated accordingly, so that all Kubernetes minor versions supported by RKE consume the add-on version that includes the commit
```
Could you elaborate more about the `RKE k8s version backports` case? I am probably missing that one but I'm not sure where is the commit in that case. Thanks Seb!

Reply:
I think this section needs to be divided in categories of changes, that we can identify and refer to in a way that is clear enough that we all know what is what. Sentences like `A commit should be available in all active Kubernetes minor versions` do not clarify enough on what the goal is.

This needs a bit of thinking to get everything cleared up, I've started a bit below:

- RKE code change is, unless scoped in the code to specific Kubernetes versions, available generically. For example, better/more/less logging statements can be merged and will be available to every invocation of RKE. Kubernetes versions are not related in this way because new RKE code gets executed wether you have a current-2, current-1 or current patch and or minor version). Backporting this means copying the code from `release/v1.X` to `release/v1.X-1`, and will be available to every invocation of the new RKE code for `release/v1.X-1`.
- RKE code changes that are scoped to Kubernetes versions/Kubernetes component versions. See https://github.com/rancher/rke/blob/645e80840596a8616c2a1f35cea0c655ac700fa9/cluster/plan.go#L60 as an example. This is outside of KDM.

Both changes above should be backported if they are related to versions that are supported in both `release/v1.X` and `release/v1.X-1` because we don't want users to have differences between provisioning a cluster in `release/v1.X-1` or `release/v1.X`. (If a flag is changed between versions that is only available in one version, an edit to the cluster will trigger provisioning (makes sense) but it will then also replace the component with the newly added flag to be removed if its not in `release/v1.X` (or vice versa)

This is just a beginning, and we don't need to flesh it out completely but this is the basics that we need to nail down to have a starting point. I am also fine with starting with what we have currently but I don't think its a good enough starting point, I'd rather take it a step further and start with more than this.

I also prefer having more clear statements than `By active we mean minor versions that Kubernetes upstream is actively working on, which normally means 3 kubernetes minors.`. We should make a statement now and say that we do latest supported - 2 for example. That is already way more clear than the previous statement. We can fine tune from there but we need some more clarity on these topics.

Lastly, the way things are described need to be more clear. Sentence like `However, if the commit is done in a dependent project, such as an add-on, the `data.json` file will need to be updated accordingly,` is weird because we don't update `data.json`. We updated source files in KDM and generate the `data.json`. I know this sounds nit picky, but for me its the difference between having a lot of confusion between multiple readers and having clarity (saving time because its more clear)

Hope it makes sense and let me know if you get what I mean or not, we can also have a session to go over it.

Reply:
Thanks for the explanation. My intention was not to cover all the different options in detail as then the document will probably become too long. My experience tells me that too long documents have the danger that nobody ever reads them, thus making this PR a bit useless. But let me try to cover most or all backport use cases and see what we get

Reply:
I made an update going more into detail and with example. The result is not as long as I anticipated :)